### PR TITLE
OT for annotation ops

### PR DESCRIPTION
**Pull request by kossebau**
If in March I could no longer see XML structures, the reason are these commits... ;)
6,815 additions in webodf/tests/ops/transformationtests.xml... manually done, and these are only what is left over...

I am sure we soon need another level of abstraction to generate those test data xml. So much duplication there with only little variants.
More serious, the transformation tests also have a problem: they do not test those part of the op specs which are just there for OT, but not used during op execution. No idea yet how to extend the test for them, not started thinking about that, just mentioning here already.

Back to the topic, this patch extends OT support to all the annotation ops and should be at least on par with the quality of the existing OT code. Actually the tests cover quite a lot of corner cases, so WRT the execution part of the opspec I feel well. 

Manual testing also showed it working well already. So please give it some good look, it is a perfect read for a relaxed sunny evening outside with your laptop :smile: really!1!

Looking forward to have Bella (& Bello) soon giving it some more stress, once also part of master (which I will be working on in the next days).

### CODE REVIEW

**Comment by kossebau** [comment]
test this please

**Comment by adityab** [comment]
I have been reviewing this for past 3.5 hours straight and after a while my eyes started glossing over the XML. But wow, great work! I don't seem to have any issues with the code so far. I'm now going to proceed to do a lot of manual testing.

File: webodf/lib/ops/OperationTransformMatrix.js [hunk 1]
```diff
@@ -274,6 +274,262 @@ ops.OperationTransformMatrix = function OperationTransformMatrix() {
...
+
+        if (addAnnotationSpecA.position < addAnnotationSpecB.position) {
+            aBeforeB = true;
+        } else if (addAnnotationSpecB.position < addAnnotationSpecA.position) {
+            aBeforeB = false;
+        } else {
+            aBeforeB = hasAPriority;
+        }
+
+        if (aBeforeB) {
```
Comment (peitschie) [inline webodf/lib/ops/OperationTransformMatrix.js]: If you wanted to de-dup this logic a bit, you could break it into ``firstAnnotation`` + ``secondAnnotation`` vars, and apply the same logic based on ``aBeforeB``. It's all by ref, so nothing stopping you keeping the return block the same as it is now, just prevents the duplicate lines.

File: webodf/lib/ops/OperationTransformMatrix.js [hunk 2]
```diff
@@ -274,6 +274,262 @@ ops.OperationTransformMatrix = function OperationTransformMatrix() {
...
+        };
+    }
+
+    /**
+     * @param {!ops.OpAddAnnotation.Spec} addAnnotationSpec
+     * @param {!ops.OpInsertText.Spec} insertTextSpec
+     * @return {?{opSpecsA:!Array.<!Object>, opSpecsB:!Array.<!Object>}}
+     */
+    function transformAddAnnotationInsertText(addAnnotationSpec, insertTextSpec) {
+        if (insertTextSpec.position <= addAnnotationSpec.position) {
```
Comment (peitschie) [inline webodf/lib/ops/OperationTransformMatrix.js]: Part of me wants to wrap this up in a standard function, as it's an often repeated pattern here, and having it common would allow some easier high-level checks.

Possibly ``moveOrExpandRegions``? Would make reviewing easier too!

Comment (peitschie) [inline webodf/lib/ops/OperationTransformMatrix.js]: Could even expand the operations that should be transformed like this to have a common interface (E.g., getRegion, insertStepsBefore, insertStepsWithin)

Comment (peitschie) [inline webodf/lib/ops/OperationTransformMatrix.js]: Hrmm... not sure still looking further. Oh well.. will keep pondering.

File: webodf/lib/ops/OperationTransformMatrix.js [hunk 3]
```diff
@@ -274,6 +274,262 @@ ops.OperationTransformMatrix = function OperationTransformMatrix() {
...
+            opSpecsB:  [moveCursorSpec]
+        };
+    }
+
+    /**
+     * @param {!ops.OpAddAnnotation.Spec} addAnnotationSpec
+     * @param {!ops.OpRemoveAnnotation.Spec} removeAnnotationSpec
+     * @return {?{opSpecsA:!Array.<!Object>, opSpecsB:!Array.<!Object>}}
+     */
+    function transformAddAnnotationRemoveAnnotation(addAnnotationSpec, removeAnnotationSpec) {
```
Comment (peitschie) [inline webodf/lib/ops/OperationTransformMatrix.js]: Logic is very similar to ``transformAddAnnotationInsertText``, just inverted. The ``+ 2`` constant is due to the fact that an annotation has a base length of 2 that's effectively unreported.

**Comment by peitschie** [comment]
Overall, this looks fine to me. I haven't checked the tests thoroughly, but every time I've wanted to see a test to prove your logic is correct, it's been there :+1: 

One thing that really strikes me is that we haven't progressed the state-of-the-art here much at all since the really early days. The fact that we're this far along and still writing hand-rolled transforms for each op pair is quite painful. Ideally, it seems like we should be working towards some higher level concepts that can be reused (e.g., multi-step add, point insert, region overlap etc.).

File: webodf/lib/ops/OperationTransformMatrix.js (hunk 1 above)
Comment (kossebau) [inline webodf/lib/ops/OperationTransformMatrix.js]: Agreed, avoiding duplicated logic helps in general. Done.

File: webodf/lib/ops/OperationTransformMatrix.js (hunk 3 above)
Comment (kossebau) [inline webodf/lib/ops/OperationTransformMatrix.js]: I tried to distill an action proposal out of this, but failed. Because it was in a homeopathic dose? :)

File: webodf/lib/ops/OperationTransformMatrix.js (hunk 2 above)
Comment (kossebau) [inline webodf/lib/ops/OperationTransformMatrix.js]: Yes, having this code less magic and more higher level would be nice. No golden bullet seen in shops so far :/

**Comment by kossebau** [comment]
Thanks @adityab and @peitschie for having taken your brain to the pain of reading all this code. Currently waiting for some more feedback, so if you were planning some, take your time, no pressing need to have this in this week. But next week, without further comments, I would then be tempted to merge :)

> One thing that really strikes me is that we haven't progressed the state-of-the-art here much at all since the really early days. The fact that we're this far along and still writing hand-rolled transforms for each op pair is quite painful. Ideally, it seems like we should be working towards some higher level concepts that can be reused (e.g., multi-step add, point insert, region overlap etc.).

Yes, as you can see in my initial comment I completely agree.

**Comment by peitschie** [comment]
I'm happy for this to :ship: whenever you're ready. The code is definitely an improvement either way :smile: 

@adityab do you have further feedback?

Also, I've started hacking up a wiki page (https://github.com/kogmbh/WebODF/wiki/Operation-Transform-Concepts) to try and tease out some of these concepts. Even just framing the cases in a structured fashion would help the code immensely, as currently each individual transform is rather unique depending on whether the dev started from ``opA`` or ``opB``, and which cases they thought of first.

**Comment by kossebau** [comment]
> Good lord. Thank you. Why did we not do this earlier? :)

No need to call me Lord, Sir @adityab :) I guess because it resulted in more lines of code, so we could praise ourselves for more lines of tests (like I just did in this PR ;) )

> I'm now going to proceed to do a lot of manual testing.

Still noting down all the issues you found? Or too afraid to tell what you saw? Or... worked good enough to have this in?

**Comment by kossebau** [comment]
> Also, I've started hacking up a wiki page (https://github.com/kogmbh/WebODF/wiki/Operation-Transform-Concepts) to try and tease out some of these concepts.

:+1: Good to note the ideas when you have them. No time myself to add to them right now, but looking forward what you might come up with so far :)

Myself hoping to push 428 and 866 closer to master again in the coming days (well, weeks)...

**Comment by adityab** [comment]
Please merge!

**Comment by kossebau** [comment]
@adityab Thanks, so will do :)
